ideascube/serveradmin/catalog.py

```diff
@@ -0,0 +1,507 @@
+
+                book = books[0]
+                book.set('path', 'data/content/%s' % zimname)
+                book.set('indexPath', 'data/index/%s.idx' % zimname)
+
+                library.append(book)
+
+        with open(os.path.join(self._install_dir, 'library.xml'), 'wb') as f:
+            f.write(etree.tostring(
+                library, xml_declaration=True, encoding='utf-8'))
+
+        try:
+            manager = SystemManager()
+            kiwix = manager.get_service('kiwix-server')
+
+        except NoSuchUnit:
+            # Kiwix is not installed, give up
+            pass
+
+        else:
+            manager.restart(kiwix.Id)
+
+
+class Catalog:
+    def __init__(self):
+        self._cache_base_dir = settings.CATALOG_CACHE_BASE_DIR
+        os.makedirs(self._cache_base_dir, exist_ok=True)
+
+        self._cache_remote_dir = os.path.join(self._cache_base_dir, 'remotes')
+        self._load_remotes()
+
+        self._cache_catalog = os.path.join(self._cache_base_dir, 'catalog.yml')
+        self._local_package_cache = os.path.join(
+            self._cache_base_dir, 'packages')
+
+        self._load_cache()
+
+    def _progress(self, msg, i, chunk_size, remote_size):
+        percent_done = (i + 1) * chunk_size / remote_size
+
+        if percent_done > 1.0:
+            percent_done = 1.0
+
+        length = 79 - len(msg) - 2 - 1 - 6
+        done_chars = int(percent_done * length)
+        remain_chars = length - done_chars
+        percent_done = int(percent_done * 1000) / 10
+
+        p = '\r{}: {}{} {}%'.format(
+            msg, "█" * done_chars, " " * remain_chars, percent_done)
+        sys.stdout.write(p)
+
+        if percent_done == 100.0:
+            sys.stdout.write('\n')
+
+        sys.stdout.flush()
+
+    # -- Manage packages ------------------------------------------------------
+    def _get_package(self, id, source):
+        try:
+            metadata = source[id]
+
+        except KeyError:
+            raise NoSuchPackage(id)
+
+        try:
+            type = metadata['type']
+
+        except KeyError:
+            raise InvalidPackageMetadata('Packages must have a type')
+
+        try:
+            return Package.registered_types[type](id, metadata)
+
+        except KeyError:
+            raise InvalidPackageType(type)
+
+    def _get_packages(self, id_patterns, source, fail_if_no_match=True):
+        pkgs = []
+
+        for id_pattern in id_patterns:
+            if '*' in id_pattern:
+                for id in source:
+                    if fnmatch(id, id_pattern):
+                        pkgs.append(self._get_package(id, source))
+
+            else:
+                try:
+                    pkgs.append(self._get_package(id_pattern, source))
+
+                except NoSuchPackage as e:
+                    if fail_if_no_match:
+                        raise e
+
+        return pkgs
+
+    def _get_handler(self, package):
+        try:
+            handlertype = package.handler
```

Comment:
But then how do you know which handler to use for a given package?

Originally we had decided to separate the `type` and `handler` [1] package metadata because a given handler might be able to handle various package types, and a given package type might be handled by various handler, depending on the actual package.

---

[1] which we used to call `app`, but since there's also an `app` package type I found it confusing and renamed it